docs/examples/expressions/expr_type_promotions.ipynb

```diff
@@ -0,0 +1,416 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Prepare the test pipeline\n",
+    "\n",
+    "First, we will prepare helper code, so we can easily manipulate the types and values that will appear as tensors in DALI pipeline.\n",
+    "\n",
+    "We use `from __future__ import division` to allow `/` and `//` as true division and floor division operators.\n",
+    "We will be using numpy as source for the custom provided data and we also need to import several things from DALI needed to create Pipeline and use ExternalSource Operator. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from __future__ import division\n",
+    "import numpy as np\n",
+    "from nvidia.dali.pipeline import Pipeline\n",
+    "import nvidia.dali.ops as ops            \n",
+    "import nvidia.dali.types as types\n",
+    "from nvidia.dali.types import Constant\n",
+    "\n",
+    "batch_size = 1"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Defining the data\n",
+    "\n",
+    "As we are dealing with binary operators, we need two inputs. \n",
+    "We will create a simple helper function that returns two numpy arrays of given numpy types with arbitrary selected values. It is to make the manipulation of types easy. In an actual scenario the data processed by DALI arithmetic operators would be tensors produced by other Operator containing some images, video sequences or other data.\n",
+    "\n",
+    "Keep in mind that shapes of both inputs need to match as those will be elementwise operations. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "left_magic_values = [42, 8]\n",
+    "right_magic_values = [9, 2]\n",
+    "\n",
+    "def get_data(left_type, right_type):\n",
+    "    return ([left_type(left_magic_values)], [right_type(right_magic_values)])\n",
+    "\n",
+    "batch_size = 1"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Defining the pipeline\n",
+    "\n",
+    "The next step is to define the Pipeline. We override `Pipeline.iter_setup`, a method called by the pipeline before every `Pipeline.run`. It is meant to feed the data into `ExternalSource()` operators indicated by `self.left` and `self.right`.\n",
+    "The data will be obtained from `get_data` function to which we pass the left and right types. \n",
+    "\n",
+    "Note, that we do not need to instantiate any additional operators, we can use regular Python arithmetic expression on the results of other operators in the `define_graph` step.\n",
+    "\n",
+    "For convenience, we'll wrap the usage of arithmetic operations in a lambda called `operation`, specified when creating the pipeline.\n",
```

Comment:
While lambda provides an easy way to parameterize example it only obfuscates the fact that you can just call arithmetic op on the graph edges (it is shown in expr_blend_image.ipynb but one needs to start reading examples from it).
Maybe in the top paragraph, you can point to a more basic example in `expr_blend_image.ipynb` and say that this one shows more edge cases and is rather an advanced tutorial?

Reply:
I split it into two tutorials, the problem is the data setup is a bit repetitive now.